README.md
[![Build Status](https://travis-ci.org/moznion/Time-Duration-Abbreviated.png?branch=master)](https://travis-ci.org/moznion/Time-Duration-Abbreviated) [![Coverage Status](https://coveralls.io/repos/moznion/Time-Duration-Abbreviated/badge.png?branch=master)](https://coveralls.io/r/moznion/Time-Duration-Abbreviated?branch=master)
# NAME

Time::Duration::Abbreviated - Describe time duration in abbreviated English

# SYNOPSIS

    use Time::Duration::Abbreviated;

    print duration(12345, 2); # 3 hrs 26 min
    print earlier(12345, 2);  # 3 hrs 26 min ago
    print later(12345, 2);    # 3 hrs 26 min later

    print duration_exact(12345); # 3 hrs 25 min 45 sec
    print earlier_exact(12345);  # 3 hrs 25 min 45 sec ago
    print later_exact(12345);    # 3 hrs 25 min 45 sec later

# DESCRIPTION

Time::Duration::Abbreviated is a abbreviated version of [Time::Duration](https://metacpan.org/pod/Time::Duration).

# SEE ALSO

[Time::Duration](https://metacpan.org/pod/Time::Duration)

# LICENSE

Copyright (C) moznion.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

# AUTHOR

moznion <moznion@gmail.com>


Changes
Revision history for Perl extension Time-Duration-Abbreviated

{{$NEXT}}

0.01 2014-04-21T02:01:53Z

    - original version



Build.PL
# =========================================================================
# THIS FILE IS AUTOMATICALLY GENERATED BY MINILLA.
# DO NOT EDIT DIRECTLY.
# =========================================================================

use 5.008_001;

use strict;
use warnings;
use utf8;

use Module::Build;
use File::Basename;
use File::Spec;
use CPAN::Meta;
use CPAN::Meta::Prereqs;

my %args = (
    license              => 'perl',
    dynamic_config       => 0,

    configure_requires => {
        'Module::Build' => 0.38,
    },

    name            => 'Time-Duration-Concise',
    module_name     => 'Time::Duration::Concise',
    allow_pureperl => 0,

    script_files => [glob('script/*'), glob('bin/*')],
    c_source     => [qw()],
    PL_files => {},

    test_files           => ((-d '.git' || $ENV{RELEASE_TESTING}) && -d 'xt') ? 't/ xt/' : 't/',
    recursive_test_files => 1,

    
);
if (-d 'share') {
    $args{share_dir} = 'share';
}

my $builder = Module::Build->subclass(
    class => 'MyBuilder',
    code => q{
        sub ACTION_distmeta {
            die "Do not run distmeta. Install Minilla and `minil install` instead.\n";
        }
        sub ACTION_installdeps {
            die "Do not run installdeps. Run `cpanm --installdeps .` instead.\n";
        }
    }
)->new(%args);
$builder->create_build_script();

my $mbmeta = CPAN::Meta->load_file('MYMETA.json');
my $meta = CPAN::Meta->load_file('META.json');
my $prereqs_hash = CPAN::Meta::Prereqs->new(
    $meta->prereqs
)->with_merged_prereqs(
    CPAN::Meta::Prereqs->new($mbmeta->prereqs)
)->as_string_hash;
my $mymeta = CPAN::Meta->new(
    {
        %{$meta->as_struct},
        prereqs => $prereqs_hash
    }
);
print "Merging cpanfile prereqs to MYMETA.yml\n";
$mymeta->save('MYMETA.yml', { version => 1.4 });
print "Merging cpanfile prereqs to MYMETA.json\n";
$mymeta->save('MYMETA.json', { version => 2 });


lib/Time/Duration/Concise.pm
package Time::Duration::Concise;
use 5.008005;
use strict;
use warnings;
use Time::Duration qw();
use parent qw(Exporter);

our $VERSION = "0.01";

our @EXPORT = qw(later later_exact earlier earlier_exact
                 ago ago_exact from_now from_now_exact
                 duration duration_exact concise);
our @EXPORT_OK = ('interval', @EXPORT);

sub concise {
    Time::Duration::concise($_[0]);
}

sub later {
    interval($_[0], $_[1], '%s ago', '%s later');
}

sub later_exact {
    interval_exact($_[0], '%s ago', '%s later');
}

sub earlier {
    interval($_[0], $_[1], '%s later', '%s ago');
}

sub earlier_exact {
    interval_exact($_[0], '%s later', '%s ago');
}

sub ago {
    &earlier
}

sub ago_exact {
    &earlier_exact
}

sub from_now {
    &later
}

sub from_now_exact {
    &later_exact
}

sub duration_exact {
    (my $span = shift) || return '0 sec';
    _render('%s', Time::Duration::_separate(abs $span));
}

sub duration {
    (my $span = shift) || return '0 sec';
    my $precision = int(shift || 0) || 2;  # precision (default: 2)
    _render(
        '%s',
        Time::Duration::_approximate($precision, Time::Duration::_separate(abs $span))
    );
}

sub interval_exact {
    my ($span, $neg_direction, $pos_direction) = @_;

    _render(
        _determine_direction($span, $neg_direction, $pos_direction),
        Time::Duration::_separate($span)
    );
}

sub interval {
    my ($span, $precision, $neg_direction, $pos_direction) = @_;

    $precision = int($precision || 0) || 2;
    _render(
        _determine_direction($span, $neg_direction, $pos_direction),
        Time::Duration::_approximate($precision, Time::Duration::_separate($span))
    );
}

sub _determine_direction {
    my ($span, $neg_direction, $pos_direction) = @_;

    no warnings qw(numeric uninitialized);
    my $direction = ($span <= -1) ? $neg_direction
                  : ($span >=  1) ? $pos_direction
                  : 'now';
    use warnings;

    return $direction;
}

my %units = (
    second => 'sec',
    minute => 'min',
    hour   => 'hr',
    day    => 'day',
    year   => 'yr',
);

sub _render {
    my ($direction, @pieces) = @_;

    my @wheel;
    for my $piece (@pieces) {
        next if $piece->[1] == 0;

        my $val  = $piece->[1];
        my $unit = $units{$piece->[0]};
        if ($unit =~ /\A(?:hr|day|yr)\Z/) {
            $unit .= 's' if $val > 1;
        }

        push @wheel, "$val $unit";
    }

    return "now" unless @wheel;
    return sprintf($direction, join ' ', @wheel);
}

1;
__END__

=encoding utf-8

=head1 NAME

Time::Duration::Concise - It's new $module

=head1 SYNOPSIS

    use Time::Duration::Concise;

=head1 DESCRIPTION

Time::Duration::Concise is ...

=head1 LICENSE

Copyright (C) moznion.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

moznion E<lt>moznion@gmail.comE<gt>

=cut



lib/Time/Duration/Abbreviated.pm
package Time::Duration::Abbreviated;
use 5.008005;
use strict;
use warnings;
use Time::Duration qw();
use parent qw(Exporter);

our $VERSION = "0.01";

our @EXPORT = qw(later later_exact earlier earlier_exact
                 ago ago_exact from_now from_now_exact
                 duration duration_exact concise);
our @EXPORT_OK = ('interval', @EXPORT);

sub concise {
    Time::Duration::concise($_[0]);
}

sub later {
    interval($_[0], $_[1], '%s ago', '%s later');
}

sub later_exact {
    interval_exact($_[0], '%s ago', '%s later');
}

sub earlier {
    interval($_[0], $_[1], '%s later', '%s ago');
}

sub earlier_exact {
    interval_exact($_[0], '%s later', '%s ago');
}

sub ago {
    &earlier
}

sub ago_exact {
    &earlier_exact
}

sub from_now {
    &later
}

sub from_now_exact {
    &later_exact
}

sub duration_exact {
    (my $span = shift) || return '0 sec';
    _render('%s', Time::Duration::_separate(abs $span));
}

sub duration {
    (my $span = shift) || return '0 sec';
    my $precision = int(shift || 0) || 2;  # precision (default: 2)
    _render(
        '%s',
        Time::Duration::_approximate($precision, Time::Duration::_separate(abs $span))
    );
}

sub interval_exact {
    my ($span, $neg_direction, $pos_direction) = @_;

    _render(
        _determine_direction($span, $neg_direction, $pos_direction),
        Time::Duration::_separate($span)
    );
}

sub interval {
    my ($span, $precision, $neg_direction, $pos_direction) = @_;

    $precision = int($precision || 0) || 2;
    _render(
        _determine_direction($span, $neg_direction, $pos_direction),
        Time::Duration::_approximate($precision, Time::Duration::_separate($span))
    );
}

sub _determine_direction {
    my ($span, $neg_direction, $pos_direction) = @_;

    no warnings qw(numeric uninitialized);
    my $direction = ($span <= -1) ? $neg_direction
                  : ($span >=  1) ? $pos_direction
                  : 'now';
    use warnings;

    return $direction;
}

my %units = (
    second => 'sec',
    minute => 'min',
    hour   => 'hr',
    day    => 'day',
    year   => 'yr',
);

sub _render {
    my ($direction, @pieces) = @_;

    my @wheel;
    for my $piece (@pieces) {
        next if $piece->[1] == 0;

        my $val  = $piece->[1];
        my $unit = $units{$piece->[0]};
        if ($unit =~ /\A(?:hr|day|yr)\Z/) {
            $unit .= 's' if $val > 1;
        }

        push @wheel, "$val $unit";
    }

    return "now" unless @wheel;
    return sprintf($direction, join ' ', @wheel);
}

1;
__END__

=encoding utf-8

=head1 NAME

Time::Duration::Abbreviated - Describe time duration in abbreviated English

=head1 SYNOPSIS

    use Time::Duration::Abbreviated;

    print duration(12345, 2); # 3 hrs 26 min
    print earlier(12345, 2);  # 3 hrs 26 min ago
    print later(12345, 2);    # 3 hrs 26 min later

    print duration_exact(12345); # 3 hrs 25 min 45 sec
    print earlier_exact(12345);  # 3 hrs 25 min 45 sec ago
    print later_exact(12345);    # 3 hrs 25 min 45 sec later

=head1 DESCRIPTION

Time::Duration::Abbreviated is a abbreviated version of L<Time::Duration>.

=head1 SEE ALSO

L<Time::Duration>

=head1 LICENSE

Copyright (C) moznion.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

moznion E<lt>moznion@gmail.comE<gt>

=cut

